File: Lab 1/frontendB2B/src/services/reporteVentasServicio.ts
import api from '@/http-common'

/**
 * Servicio para consultar reportes de ventas desde el backend
 * Utiliza la vista materializada vw_ventas_mensuales_por_categoria
 */
export const reporteVentasServicio = {
  
  /**
   * Obtiene todos los reportes de ventas mensuales consolidados por categoría
   * @returns Lista de reportes con mes/año, categoría, cantidad de órdenes y total
   */
  async obtenerTodosLosReportes() {
    try {
      const response = await api.get('/reportes/ventas')
      return response.data
    } catch (error) {
      console.error('Error al obtener reportes:', error)
      throw error
    }
  },

  /**
   * Obtiene reportes de un mes específico
   * @param mesAno Formato YYYY-MM (ej: 2024-03)
   * @returns Reportes filtrados por mes y año
   */
  async obtenerPorMesAno(mesAno: string) {
    try {
      const response = await api.get('/reportes/ventas/mes', {
        params: { mesAno }
      })
      return response.data
    } catch (error) {
      console.error('Error al obtener reportes por mes:', error)
      throw error
    }
  },

  /**
   * Obtiene reportes de una categoría específica
   * @param nombre Nombre de la categoría
   * @returns Reportes filtrados por categoría
   */
  async obtenerPorCategoria(nombre: string) {
    try {
      const response = await api.get('/reportes/ventas/categoria', {
        params: { nombre }
      })
      return response.data
    } catch (error) {
      console.error('Error al obtener reportes por categoría:', error)
      throw error
    }
  },

  /**
   * Obtiene reportes de un año específico
   * @param anio Año (ej: 2024)
   * @returns Reportes filtrados por año
   */
  async obtenerPorAnio(anio: number) {
    try {
      const response = await api.get('/reportes/ventas/anio', {
        params: { anio }
      })
      return response.data
    } catch (error) {
      console.error('Error al obtener reportes por año:', error)
      throw error
    }
  },

  /**
   * Obtiene el total consolidado de todas las ventas
   * Útil para KPIs y métricas generales
   * @returns Objeto con totales: cantidadOrdenes, cantidadProductos, totalVendido, precioPromedio
   */
  async obtenerTotalConsolidado() {
    try {
      const response = await api.get('/reportes/ventas/total')
      return response.data
    } catch (error) {
      console.error('Error al obtener total consolidado:', error)
      throw error
    }
  },

  /**
   * Refresca la vista materializada (endpoint administrativo)
   * Ejecutar después de cambios significativos en órdenes o productos
   */
  async refrescarReportes() {
    try {
      const response = await api.post('/reportes/refrescar')
      return response.data
    } catch (error) {
      console.error('Error al refrescar reportes:', error)
      throw error
    }
  }
}

/**
 * Tipo de dato para los reportes
 */
export interface ReporteVentas {
  mesAno: string              // Formato: YYYY-MM
  anio: number               // Año
  mes: number                // Mes (1-12)
  nombreCategoria: string    // Nombre de la categoría
  cantidadOrdenes: number    // Cantidad de órdenes en el mes
  cantidadProductos: number  // Total de productos vendidos
  totalVendido: number       // Total de dinero vendido
  precioPromedio: number     // Precio promedio
}

File: Lab 3/backendB2B/src/main/java/com/ecommerceb2b/backend/Services/ReporteVentasServicio.java
package com.ecommerceb2b.backend.Services;

import com.ecommerceb2b.backend.Entities.ReporteVentasEntidad;
import com.ecommerceb2b.backend.Repository.ReporteVentasRepositorio;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class ReporteVentasServicio {

    private final ReporteVentasRepositorio reporteVentasRepositorio;

    public ReporteVentasServicio(ReporteVentasRepositorio reporteVentasRepositorio) {
        this.reporteVentasRepositorio = reporteVentasRepositorio;
    }

    /**
     * Obtiene todos los registros de ventas mensuales por categoría
     * Útil para el dashboard general que muestra todos los datos consolidados
     */
    @Transactional(readOnly = true)
    public List<ReporteVentasEntidad> obtenerTodosLosReportes() {
        return reporteVentasRepositorio.obtenerTodosLosReportes();
    }

    /**
     * Filtra reportes por mes y año específico (formato YYYY-MM)
     * Útil para análisis de un mes particular
     */
    @Transactional(readOnly = true)
    public List<ReporteVentasEntidad> obtenerPorMesAno(String mesAno) {
        if (mesAno == null || mesAno.trim().isEmpty()) {
            throw new IllegalArgumentException("El mes y año son obligatorios");
        }
        // Validar formato YYYY-MM
        if (!mesAno.matches("\\d{4}-\\d{2}")) {
            throw new IllegalArgumentException("El formato debe ser YYYY-MM");
        }
        return reporteVentasRepositorio.obtenerPorMesAno(mesAno);
    }

    /**
     * Filtra reportes por categoría de producto
     * Útil para análisis de desempeño por línea de producto
     */
    @Transactional(readOnly = true)
    public List<ReporteVentasEntidad> obtenerPorCategoria(String nombreCategoria) {
        if (nombreCategoria == null || nombreCategoria.trim().isEmpty()) {
            throw new IllegalArgumentException("La categoría es obligatoria");
        }
        return reporteVentasRepositorio.obtenerPorCategoria(nombreCategoria.trim());
    }

    /**
     * Obtiene reportes de un año completo
     * Útil para análisis anuales
     */
    @Transactional(readOnly = true)
    public List<ReporteVentasEntidad> obtenerPorAnio(Integer anio) {
        if (anio == null || anio < 2000 || anio > 2100) {
            throw new IllegalArgumentException("El año debe estar entre 2000 y 2100");
        }
        return reporteVentasRepositorio.obtenerPorAnio(anio);
    }

    /**
     * Obtiene el total consolidado de todas las ventas
     * Útil para KPIs generales del dashboard
     */
    @Transactional(readOnly = true)
    public ReporteVentasEntidad obtenerTotalConsolidado() {
        return reporteVentasRepositorio.obtenerTotalConsolidado();
    }

    /**
     * Refresca la vista materializada para obtener datos más recientes
     * Se debe ejecutar después de cambios importantes en órdenes o productos
     */
    @Transactional
    public void refrescarReportes() {
        reporteVentasRepositorio.refrescarVistaMatrializada();
    }
}
